examples/OpenGL45/matrix_multiplication/matrix_multiplication_version7.shadertrap

```diff
@@ -0,0 +1,328 @@
+          512. 1024. 1536. 2048. 2560. 3072. 3584. 4096. 4608. 5120. 5632. 6144. 6656. 7168. 7680. 8192.
+          512. 1024. 1536. 2048. 2560. 3072. 3584. 4096. 4608. 5120. 5632. 6144. 6656. 7168. 7680. 8192.
+          512. 1024. 1536. 2048. 2560. 3072. 3584. 4096. 4608. 5120. 5632. 6144. 6656. 7168. 7680. 8192.
+          512. 1024. 1536. 2048. 2560. 3072. 3584. 4096. 4608. 5120. 5632. 6144. 6656. 7168. 7680. 8192.
+    
+DECLARE_SHADER array_to_blocked_array KIND COMPUTE
+#version 450
+
+#define KERNEL_SIZE 4
+#define MAT_SIZE 16
+
+layout(local_size_x=1, local_size_y=1, local_size_z=1) in;
+
+layout(std430, binding = 0) buffer array1 {
+  float data[MAT_SIZE][MAT_SIZE];
+};
+layout(std430, binding=1) buffer array1_mat_buf{
+  mat4x4 array_mat[MAT_SIZE/KERNEL_SIZE][MAT_SIZE/KERNEL_SIZE];
+};
+void main() {
+    uint row = uint(gl_WorkGroupID[0]), col = uint(gl_WorkGroupID[1]);
+    array_mat[row/KERNEL_SIZE][col/KERNEL_SIZE][col%KERNEL_SIZE][row%KERNEL_SIZE] = data[row][col];
+}
+END
+
+COMPILE_SHADER array_to_blocked_array_compiled SHADER array_to_blocked_array
+CREATE_PROGRAM array_to_blocked_array_prog SHADERS array_to_blocked_array_compiled
+
+BIND_SHADER_STORAGE_BUFFER BUFFER array1 BINDING 0
+BIND_SHADER_STORAGE_BUFFER BUFFER array1_mat_buf BINDING 1
+
+RUN_COMPUTE
+    PROGRAM array_to_blocked_array_prog
+    NUM_GROUPS 16 16 1
+
+ASSERT_EQUAL BUFFERS array1_mat_buf correct_array1_mat_buf
+
+BIND_SHADER_STORAGE_BUFFER BUFFER array2 BINDING 0
+BIND_SHADER_STORAGE_BUFFER BUFFER array2_mat_buf BINDING 1
+
+RUN_COMPUTE
+    PROGRAM array_to_blocked_array_prog
+    NUM_GROUPS 16 16 1
+
+ASSERT_EQUAL BUFFERS array2_mat_buf correct_array2_mat_buf
+
+DECLARE_SHADER mat_mul KIND COMPUTE
+#version 450
+
+#define KERNEL_SIZE 4
+#define MAT_SIZE 16
+
+layout(local_size_x=4, local_size_y=4, local_size_z=1) in;
+
+layout(std430, binding=0) buffer array1_mat_buf{
+  mat4x4 array1_mat[MAT_SIZE/KERNEL_SIZE][MAT_SIZE/KERNEL_SIZE];
+};
+layout(std430, binding=1) buffer array2_mat_buf{
+  mat4x4 array2_mat[MAT_SIZE/KERNEL_SIZE][MAT_SIZE/KERNEL_SIZE];
+};
+layout(std430, binding=2) buffer blocked_result{
+  mat4x4 blocked_res[MAT_SIZE/KERNEL_SIZE][MAT_SIZE/KERNEL_SIZE];
+};
+
+shared mat4x4 data1_sub, data2_sub;
+void main() {    
+    uint row = uint(gl_WorkGroupID[0]), col = uint(gl_WorkGroupID[1]);
+    uint x = uint(gl_LocalInvocationID[0]), y = uint(gl_LocalInvocationID[1]);
+    blocked_res[row][col] = mat4x4(0);
+    
+    for(int i = 0; i < MAT_SIZE/KERNEL_SIZE; i++){
+
+        // Each thread reads one element of the submatrix
+        data1_sub[x][y] = array1_mat[row][i][x][y];
+        data2_sub[x][y] = array2_mat[i][col][x][y];
+        
+        // Wait until each thread has read its corresponding element         
+        barrier();
+
+        // Each thread writes to one element of the resulting matrix
+        blocked_res[row][col] += (data1_sub * data2_sub);
+        barrier();
+    }
```

Comment:
Hmm, I don't think this is a sensible way to use shared memory.

I was imagining a tiled matrix multiplication algorithm that uses shared memory per tile. Could you look into that?

Cheers

Ally

Reply:
@afd Yes, I had the exact same thing in mind, but I think that the code does that, no?
It first splits the array into 4x4 submatrices and then uses a shared memory for each element of the submatrix.
I based my approach on the following:
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#shared-memory

Reply:
OK, I think I get it - you're using a sub-matrix as a 4x4 tile? If so then that makes sense.

Reply:
Yes, exactly.